src/modules/flight_mode_manager/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp

```diff
@@ -218,12 +218,26 @@ void FlightTaskAutoLineSmoothVel::_generateSetpoints()
 void FlightTaskAutoLineSmoothVel::_checkEmergencyBraking()
 {
 	if (!_is_emergency_braking_active) {
-		if (_position_smoothing.getCurrentVelocityZ() > (2.f * _param_mpc_z_vel_max_dn.get())) {
+		// activate emergency braking if significantly outside of velocity bounds
+		const float factor = 1.3f;
+		const bool is_vertical_speed_exceeded = _position_smoothing.getCurrentVelocityZ() >
+							(factor * _param_mpc_z_vel_max_dn.get())
+							|| _position_smoothing.getCurrentVelocityZ() < -(factor * _param_mpc_z_vel_max_up.get());
+		const bool is_horizontal_speed_exceeded = _position_smoothing.getCurrentVelocityXY().longerThan(
+					factor * _param_mpc_xy_cruise.get());
+
+		if (is_vertical_speed_exceeded || is_horizontal_speed_exceeded) {
 			_is_emergency_braking_active = true;
 		}
 
 	} else {
-		if (fabsf(_position_smoothing.getCurrentVelocityZ()) < 0.01f) {
+		// deactivate emergency braking if all velocities and accelerations are again within their bounds
+		if (_position_smoothing.getCurrentVelocityZ() < _param_mpc_z_vel_max_dn.get()
```

Comment:
```suggestion
		if (_position_smoothing.getCurrentVelocityZ() < 0.01f
```

I missed that change; The reason I had this small value and not the parameter is that the time-optimal trajectory is computed assuming that the maximum jerk and acceleration is constant. If it changes dynamically, you'll end up overshooting the target velocity.

To get it properly on the XY axes, we also need to force the target velocity to be 0 while in emergency braking mode. I had that originally (https://github.com/PX4/PX4-Autopilot/pull/18277/files#diff-3fa0514afc596aefc8754a909f5ed57ce5c80429e559679c5e98e5ca7118b226R330-R331) but it got lost in a refactoring. 

To fix it, do the following:

```
void FlightTaskAutoLineSmoothVel::_generateSetpoints()
{
	_checkEmergencyBraking();
	Vector3f waypoints[] = {_prev_wp, _position_setpoint, _next_wp};

	if (isTargetModified()) {
		// In case object avoidance has injected a new setpoint, we take this as the next waypoints
		waypoints[2] = _position_setpoint;
	}

	const bool should_wait_for_yaw_align = _param_mpc_yaw_mode.get() == 4 && !_yaw_sp_aligned;
add ==>  const bool force_zero_velocity_setpoint = should_wait_for_yaw_align || _is_emergency_braking_active;
	_updateTrajConstraints();
	PositionSmoothing::PositionSmoothingSetpoints smoothed_setpoints;
	_position_smoothing.generateSetpoints(
		_position,
		waypoints,
		_velocity_setpoint,
		_deltatime,
replace ==>     force_zero_velocity_setpoint,
		smoothed_setpoints
	);

```

Reply:
Ok, so some undesired overshoots were the reason for constraining to 0 velocity. I didn't understand the reasoning behind. I didn't see why you'd want to get to a full stop if your target is still in front of you. E.g. for an RTL to closest rally point which might be ahead of the drone, so it would make more sense to keep the horizontal speed once it's within limits. But I guess that doesn't really work anyway as for now the drone always wants to go back to where the RTL was initially triggered to climb to RTL altitude before moving ahead.

Reply:
Emergency braking is not something that should occur every time, it's then fine to have a sub-optimal behavior, but it's good if there is no overshoot.